packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -729,7 +755,18 @@ class TextSelectionOverlay {
       return;
     }
 
-    _endHandleDragPosition = _getHandleDy(details.globalPosition.dy, _endHandleDragPosition);
+    // This is NOT the same as details.localPosition. That is relative to the
+    // selection handle, whereas this is relative to the RenderEditable.
+    final Offset localPosition = renderObject.globalToLocal(details.globalPosition);
+
+    final double nextEndHandleDragPositionLocal = _getHandleDy(
```

Comment:
Why do we have to deal with drag position deltas and change coordinate systems back and forth? Can we use the current location to calculate where the handle/magnifier should be?

Reply:
The trickiness here is that the handle needs to snap to the scaled lines. The snapping calculations are done in local coordinates and then it's converted back to global coordinates. After reading your comment I played around with some refactoring but ended up just renaming some things to try to make it more clear.

Reply:
Doesn't `getPositionForPoint` do the snapping? Why is additional snapping needed?

Reply:
Sorry for my bad explanation, it's more subtle than my comment made it seem. Simply maintaining the offset between the contact point and the place in the text it points to is not enough. You can see why by doing this:

 1. Get out your phone and find a multiline text field (native or Flutter app).
 2. Tap to place the cursor.
 3. Drag the cursor slowly down until it jumps to the next line.
 4. Immediately drag back up until it jumps to the original line.

Notice that it doesn't jump back immediately. There is no balance point at which the cursor can jump up or down with a slight drag, like there would be if it was simply using an offset. You must drag a full line height for it to change lines in either direction.

I'm adding some details about this in the code comments.